pkg/sql/sem/tree/type_check.go

```diff
@@ -30,8 +31,12 @@ import (
 )
 
 // SemaContext defines the context in which to perform semantic analysis on an
-// expression syntax tree.
+// expression syntax tree. It should not be constructed directly, and instead
+// should be constructed using MakeSemaContext.
 type SemaContext struct {
+	// Context is the context under which to perform semantic analysis.
+	Context context.Context
```

Comment:
(I'm sorry if we do this comment)

I'm not typically a fan out putting contexts in structs, because it can be confusing if you start introducing cancellations and the like -- https://groups.google.com/forum/#!topic/golang-nuts/xRbzq8yzKWI

I know we do this for EvalContext, but I'm not sure we should continue the precedent. In an ideal world, context is threaded through everything...

Reply:
oof. Yeah this seems reasonable. All that would really change in this PR is just typecheck itself taking a context, now that I've plumbed a context into all uses of SemaContext.